projects/fal/src/fal/api.py

```diff
@@ -873,10 +891,34 @@ def openapi(self) -> dict[str, Any]:
         return self._build_app().openapi()
 
     def serve(self) -> None:
-        import uvicorn
+        import asyncio
+
+        from starlette_exporter import handle_metrics
+        from uvicorn import Config
 
         app = self._build_app()
-        uvicorn.run(app, host="0.0.0.0", port=8080)
+        server = Server(config=Config(app, host="0.0.0.0", port=8080))
+        metrics_app = FastAPI()
+        metrics_app.add_route("/metrics", handle_metrics)
+        metrics_server = Server(config=Config(metrics_app, host="0.0.0.0", port=9090))
+
+        async def _serve() -> None:
+            event = asyncio.Event()
+            # loop = asyncio.get_running_loop()
+            # You cannot add signal handlers to any loop if you're not
+            # on the main thread.
+            # How can we detect that we are being shut down and stop the
+            # uvicorn servers gracefully?
+            # loop.add_signal_handler(signal.SIGINT, event.set)
+            # loop.add_signal_handler(signal.SIGTERM, event.set)
```

Comment:
PTAL!
The servers work, however we are not stopping gracefully AFAICT.
How can we tell that the thread is being shut down so we can proactively stop the servers nicely?

Reply:
I don't have much my own experience with this, so can't recommend anything specific. We both googled around and it seems like another general practice is to run each app in a separate thread, which makes handling this more obvious.

Reply:
cc @isidentical